File: backend/src/sandbox/api/tool/read.py
"""Internal implementation for the public sandbox file-read verb."""

from __future__ import annotations

from audit.base import AuditSink
from sandbox.api.tool._daemon_payload import daemon_request_identity
from sandbox.api.tool._operation_audit import run_audited_operation
from sandbox.api.tool._result_projection import read_result_from_daemon_response
from sandbox.api.timeouts import READ_FILE_TIMEOUT_S
from sandbox.api.transport import DAEMON_OP_READ_FILE, DaemonSandboxTransport, SandboxTransport
from sandbox._shared.models import ReadFileRequest, ReadFileResult


async def read_file(
    sandbox_id: str,
    request: ReadFileRequest,
    *,
    audit_sink: AuditSink | None = None,
    transport: SandboxTransport | None = None,
) -> ReadFileResult:
    """Read one UTF-8 text file through the sandbox daemon."""
    daemon_transport = transport or DaemonSandboxTransport()

    async def _call() -> ReadFileResult:
        payload = daemon_request_identity(request) | {"path": request.path}
        response = await daemon_transport.call(
            sandbox_id,
            DAEMON_OP_READ_FILE,
            payload,
            timeout=READ_FILE_TIMEOUT_S,
        )
        return read_result_from_daemon_response(response)

    return await run_audited_operation(
        audit_sink=audit_sink,
        sandbox_id=sandbox_id,
        operation="read_file",
        caller=request.caller,
        payload={"path": request.path},
        call=_call,
    )


__all__ = ["read_file"]

File: agent-core/crates/eos-sandbox-port/src/tool_api/read.rs
//! Pure `read_file` helper: build payload → call transport → parse envelope.

use eos_types::SandboxId;
use serde_json::Value;

use crate::error::SandboxPortError;
use crate::models::{ReadFileRequest, ReadFileResult};
use crate::ops::DaemonOp;
use crate::timeouts::READ_FILE_TIMEOUT_S;
use crate::tool_api::parse::{daemon_request_identity_fields, parse_read_file_result};
use crate::transport::SandboxTransport;

/// Read one UTF-8 text file through the sandbox daemon.
pub async fn read_file(
    transport: &dyn SandboxTransport,
    sandbox_id: &SandboxId,
    request: &ReadFileRequest,
) -> Result<ReadFileResult, SandboxPortError> {
    let mut payload = daemon_request_identity_fields(&request.base);
    payload.insert("path".to_owned(), Value::String(request.path.clone()));
    let response = transport
        .call(sandbox_id, DaemonOp::ReadFile, payload, READ_FILE_TIMEOUT_S)
        .await?;
    parse_read_file_result(&response)
}
